packages/block-editor/src/hooks/font-size.js

```diff
@@ -184,19 +163,50 @@ function useBlockProps( { name, fontSize, style } ) {
 	if (
 		! hasBlockSupport( name, FONT_SIZE_SUPPORT_KEY ) ||
 		shouldSkipSerialization( name, TYPOGRAPHY_SUPPORT_KEY, 'fontSize' ) ||
-		! fontSize ||
-		style?.typography?.fontSize
+		! fontSize
 	) {
 		return;
 	}
 
-	const fontSizeValue = getFontSize(
-		fontSizes,
-		fontSize,
-		style?.typography?.fontSize
-	).size;
+	let props = {};
+
+	if ( ! style?.typography?.fontSize ) {
+		props = {
+			style: {
+				fontSize: getFontSize(
+					fontSizes,
+					fontSize,
+					style?.typography?.fontSize
+				).size,
+			},
+		};
+	}
 
-	return { style: { fontSize: fontSizeValue } };
+	// TODO: This sucks! We should be using useSetting( 'typography.fluid' )
+	// or even useSelect( blockEditorStore ). We can't do either here
+	// because getEditWrapperProps is a plain JavaScript function called by
+	// BlockListBlock and not a React component rendered within
+	// BlockListContext.Provider. If we set fontSize using editor.
+	// BlockListBlock instead of using getEditWrapperProps then the value is
+	// clobbered when the core/style/addEditProps filter runs.
+
+	// TODO: We can do the thing above now.
+	const fluidTypographySettings = getFluidTypographyOptionsFromSettings(
+		select( blockEditorStore ).getSettings().__experimentalFeatures
+	);
+
+	if ( fontSize ) {
+		props = {
+			style: {
+				fontSize: getTypographyFontSizeValue(
+					{ size: fontSize },
+					fluidTypographySettings
+				),
```

Comment:
I'm not too sure what the right fix is, but I think this change in this PR means that the fluid typography rules are no longer calculated and output when a user enters a custom font-size within the block editor. For example, in the following screenshot, the font-size is the direct value of `121px`, but it should be getting the fluid typography rules:

<img width="1368" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/bae5b6fd-618b-4fe6-9fdc-79979b41d548">

Because this hook returns early if `fontSize` (as an attribute) isn't present, we never get to this transformation of the value when someone enters a value in `style.typography.fontSize`.

Found while testing https://github.com/WordPress/gutenberg/pull/57866#pullrequestreview-1822541005

Reply:
Thanks for spotting this @andrewserong and apologies that I didn't spot it here first.

I'm not sure what the best approach is either.

I hacked something that seems to work in principle, but the output is somehow modified before it reaches the frontend.

The fluid `fontSize` is calculated correctly, e.g., `clamp(46.987px, 2.937rem + ((1vw - 3.2px) * 4.689), 92px)`, but in the browser inspector (Chrome and FF) I'm seeing something completely different: `clamp(46.987px, 2.937rem + 4.689vw - 15.0048px, 92px);` 🤔 

Visually they look the same. I ran out of time to debug.


<details>

<summary>Diff of my hack</summary>

```diff
diff --git a/packages/block-editor/src/components/global-styles/typography-utils.js b/packages/block-editor/src/components/global-styles/typography-utils.js
index cd1340033e..a73dfc3725 100644
--- a/packages/block-editor/src/components/global-styles/typography-utils.js
+++ b/packages/block-editor/src/components/global-styles/typography-utils.js
@@ -81,7 +81,7 @@ export function getTypographyFontSizeValue( preset, typographyOptions ) {
 	return defaultSize;
 }
 
-function isFluidTypographyEnabled( typographySettings ) {
+export function isFluidTypographyEnabled( typographySettings ) {
 	const fluidSettings = typographySettings?.fluid;
 	return (
 		true === fluidSettings ||
diff --git a/packages/block-editor/src/hooks/font-size.js b/packages/block-editor/src/hooks/font-size.js
index 89491da44e..b643c2a087 100644
--- a/packages/block-editor/src/hooks/font-size.js
+++ b/packages/block-editor/src/hooks/font-size.js
@@ -26,6 +26,7 @@ import { store as blockEditorStore } from '../store';
 import {
 	getTypographyFontSizeValue,
 	getFluidTypographyOptionsFromSettings,
+	isFluidTypographyEnabled,
 } from '../components/global-styles/typography-utils';
 
 export const FONT_SIZE_SUPPORT_KEY = 'typography.fontSize';
@@ -155,33 +156,6 @@ export function useIsFontSizeDisabled( { name: blockName } = {} ) {
 
 function useBlockProps( { name, fontSize, style } ) {
 	const [ fontSizes ] = useSettings( 'typography.fontSizes' );
-
-	// Only add inline styles if the block supports font sizes,
-	// doesn't skip serialization of font sizes,
-	// doesn't already have an inline font size,
-	// and does have a class to extract the font size from.
-	if (
-		! hasBlockSupport( name, FONT_SIZE_SUPPORT_KEY ) ||
-		shouldSkipSerialization( name, TYPOGRAPHY_SUPPORT_KEY, 'fontSize' ) ||
-		! fontSize
-	) {
-		return;
-	}
-
-	let props = {};
-
-	if ( ! style?.typography?.fontSize ) {
-		props = {
-			style: {
-				fontSize: getFontSize(
-					fontSizes,
-					fontSize,
-					style?.typography?.fontSize
-				).size,
-			},
-		};
-	}
-
 	// TODO: This sucks! We should be using useSetting( 'typography.fluid' )
 	// or even useSelect( blockEditorStore ). We can't do either here
 	// because getEditWrapperProps is a plain JavaScript function called by
@@ -194,18 +168,35 @@ function useBlockProps( { name, fontSize, style } ) {
 	const fluidTypographySettings = getFluidTypographyOptionsFromSettings(
 		select( blockEditorStore ).getSettings().__experimentalFeatures
 	);
-
-	if ( fontSize ) {
-		props = {
-			style: {
-				fontSize: getTypographyFontSizeValue(
-					{ size: fontSize },
-					fluidTypographySettings
-				),
-			},
-		};
+	// Only add inline styles if the block supports font sizes,
+	// doesn't skip serialization of font sizes,
+	// doesn't already have an inline font size,
+	// and does have a class to extract the font size from.
+	if (
+		! hasBlockSupport( name, FONT_SIZE_SUPPORT_KEY ) ||
+		shouldSkipSerialization( name, TYPOGRAPHY_SUPPORT_KEY, 'fontSize' ) ||
+		( ! fontSize &&
+			! isFluidTypographyEnabled( fluidTypographySettings ) &&
+			!! style?.typography?.fontSize )
+	) {
+		return;
 	}
 
+	const props = {
+		style: {
+			fontSize: ! style?.typography?.fontSize
+				? getFontSize(
+						fontSizes,
+						fontSize,
+						style?.typography?.fontSize
+				  ).size
+				: getTypographyFontSizeValue(
+						{ size: style.typography.fontSize },
+						fluidTypographySettings
+				  ),
+		},
+	};
+
 	return addSaveProps( props, name, { fontSize } );
 }
 


```

</details>

Reply:
Should we prioritize this regression for 6.5?

Reply:
> Should we prioritize this regression for 6.5?

Yes, that'd be great if you time to. Thanks for picking this up!

Reply:
https://github.com/WordPress/gutenberg/pull/58422